### PR TITLE
Zones 1.1

### PR DESCRIPTION
**Pull request by cupOJoseph**
Creates an implementation of Zones for Seaport, including some docs, and full coverage tests. 

## Motivation

Zones let seaport orders to a pre-order hook, where cool stuff can happen.

## Solution

Creates a zones folder under contracts. the PausableZoneController can create new zones and is managed by an owner and other roles. PausableZone is a basic zone implementation anyone can use as an example; the owner of it can use it to process or cancel restricted orders, or pause order which list it as a zone. 

For more info about how it works, and what is possible.. check out the new zone documentation. A bit more work on the zone docs will be in shortly.

### CODE REVIEW

File: docs/ZoneDocumentation.md [hunk 1]
```diff
@@ -0,0 +1,41 @@
+# Zone Documentation
+
+The `zone` of the order is an optional secondary account attached to the order with two additional privileges:
+
+1. The zone may cancel orders where it is named as the zone by calling `cancel`. (Note that offerers can also cancel their own orders, either individually or for all orders signed with their current counter at once by calling `incrementCounter`).
+2. "Restricted" orders (as specified by the order type) must either be executed by the zone or the offerer, or must be approved as indicated by a call to an `isValidOrder` or `isValidOrderIncludingExtraData` view function on the zone.
+
+An example zone contract implementation can be found at `/contracts/zones/GlobalPausable.sol`
```
Comment (0age) [inline docs/ZoneDocumentation.md:8]: Renamed to “PausableZone”

File: docs/ZoneDocumentation.md [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+The `GlobalPausable` contract can be used by its deployer to cancel specific orders, execute fulfillment on restricted order, and pause all orders which use it as a zone.
+
+## Further implementation
+
+We could use something like this, to allow the owner of the zone to execute restricted orders, but it's not entirely necessary:
```
Comment (0age) [inline docs/ZoneDocumentation.md:14]: Rephrase throughout to remove the use of “we”

File: docs/ZoneDocumentation.md [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+Zones are a powerful addition to the idea of simple marketplaces. By adding additional logic to approve / reject Seaport orders, many new applications are possible. Zones could potentially be used by new marketplaces built on top of Seaport to:
+
+- Stop sales of stolen assets
+- Limit the number of NFTs from a particular collection that can be sold in a given amount of time
+- Enforce a particular floor or ceiling price for certain assets
+- Make other arbitrary calls to outside price oracles
+- Track extra incentives for users completing valid orders
+
+...and probably more that we haven't even thought of yet!
```
Comment (0age) [inline docs/ZoneDocumentation.md:39]: Too colloquial + lose “we”

File: docs/ZoneDocumentation.md [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+We encourage anyone in the world to build and deploy your own unique zones and help decentralize Seaport as a platform.
```
Comment (0age) [inline docs/ZoneDocumentation.md:41]: Same, unclear who “we” is

File: contracts/zones/PausableZone.sol [hunk 5]
```diff
@@ -0,0 +1,274 @@
...
+     * @notice Cancel an arbitrary number of orders that have agreed to use the
+     *         contract as their zone.
+     *
+     * @param seaport  The Seaport address.
+     * @param orders   The orders to cancel.
+     *
+     * @return cancelled A boolean indicating whether the supplied orders have
+     *                   been successfully cancelled.
+     */
+    function cancelOrders(
```
Comment (0age) [inline contracts/zones/PausableZone.sol:155]: Address lint warnings

File: contracts/zones/PausableZone.sol [hunk 6]
```diff
@@ -0,0 +1,274 @@
...
+     *         the contract as a zone. Restricted orders with this address as a
+     *         zone will not be fulfillable unless the zone is redeployed to the
+     *         same address.
+     */
+    function pause() external override isController {
+        // Emit an event signifying that the zone is paused.
+        emit Paused();
+
+        // Destroy the zone, sending any ether to the transaction submitter.
+        selfdestruct(payable(tx.origin));
```
Comment (0age) [inline contracts/zones/PausableZone.sol:249]: This lint rule either needs to be modified or (probably better) the behavior should be modified (maybe to take an address as an argument, or to have a configured recovery address)

File: contracts/zones/PausableZone.sol [hunk 7]
```diff
@@ -0,0 +1,274 @@
...
+     *
+     * @param operatorToAssign The address to assign as the operator.
+     */
+    function assignOperator(address operatorToAssign)
+        external
+        override
+        isController
+    {
+        // Ensure the operator being assigned is not the null address.
+        require(
```
Comment (0age) [inline contracts/zones/PausableZone.sol:263]: Replace revert strings with custom reverts

File: contracts/zones/PausableZoneController.sol [hunk 8]
```diff
@@ -0,0 +1,382 @@
...
+     *
+     * @return derivedAddress The derived address for the zone.
+     */
+    function createZone(bytes32 salt)
+        external
+        override
+        returns (address derivedAddress)
+    {
+        // Ensure the caller is the owner.
+        require(
```
Comment (0age) [inline contracts/zones/PausableZoneController.sol:87]: Replace with custom reverts

File: contracts/zones/PausableZoneController.sol [hunk 9]
```diff
@@ -0,0 +1,382 @@
...
+        PausableZone(zone).pause();
+
+        // Return a boolean indicating the pause was successful.
+        success = true;
+    }
+
+    /**
+     * @notice Cancel Seaport orders on a given zone.
+     *
+     * @param pausableZoneAddress The zone that manages the orders to be cancelled.
```
Comment (0age) [inline contracts/zones/PausableZoneController.sol:144]: Address lint warnings

File: contracts/zones/PausableZoneController.sol [hunk 10]
```diff
@@ -0,0 +1,382 @@
...
+     * @param seaportAddress      The Seaport address.
+     * @param orders              The orders to cancel.
+     */
+    function cancelOrders(
+        address pausableZoneAddress,
+        SeaportInterface seaportAddress,
+        OrderComponents[] calldata orders
+    ) external override {
+        // Ensure the caller is the owner.
+        require(
```
Comment (0age) [inline contracts/zones/PausableZoneController.sol:154]: Replace with custom revert reasons

File: contracts/zones/PausableZoneController.sol [hunk 11]
```diff
@@ -0,0 +1,382 @@
...
+        PausableZone zone = PausableZone(pausableZoneAddress);
+
+        // Call cancelOrders on the given zone.
+        zone.cancelOrders(seaportAddress, orders);
+    }
+
+    /**
+     * @notice Execute an arbitrary number of matched orders on a given zone.
+     *
+     * @param pausableZoneAddress The zone that manages the orders to be cancelled.
```
Comment (0age) [inline contracts/zones/PausableZoneController.sol:169]: Address line length lint warnings

File: contracts/zones/PausableZoneController.sol [hunk 12]
```diff
@@ -0,0 +1,382 @@
...
+     * @param seaportAddress      The Seaport address.
+     * @param orders              The orders to match.
+     * @param fulfillments        An array of elements allocating offer components
```
Comment (0age) [inline contracts/zones/PausableZoneController.sol:172]: Lint warnings

File: contracts/zones/PausableZoneController.sol [hunk 13]
```diff
@@ -0,0 +1,382 @@
...
+     *                    orders.
+     */
+    function executeMatchOrders(
+        address pausableZoneAddress,
+        SeaportInterface seaportAddress,
+        Order[] calldata orders,
+        Fulfillment[] calldata fulfillments
+    ) external payable override returns (Execution[] memory executions) {
+        // Ensure the caller is the owner.
+        require(
```
Comment (0age) [inline contracts/zones/PausableZoneController.sol:186]: Use custom revert

File: contracts/zones/interfaces/PausableZoneControllerInterface.sol [hunk 14]
```diff
@@ -0,0 +1,175 @@
...
+    ) external;
+
+    /**
+     * @notice Execute an arbitrary number of matched orders on a given zone.
+     *
+     * @param pausableZoneAddress The zone that manages the orders to be
+     *                            cancelled.
+     * @param seaportAddress      The Seaport address.
+     * @param orders              The orders to match.
+     * @param fulfillments        An array of elements allocating offer components
```
Comment (0age) [inline contracts/zones/interfaces/PausableZoneControllerInterface.sol:69]: Line length

File: docs/ZoneDocumentation.md [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+    require(msg.sender == deployerOwner, "Only the owner can execute orders with the zone.");
+
+    GlobalPausable gp = GlobalPausable(_globalPausableAddress);
+    gp.executeRestrictedOffer(_seaportAddress, orders);
+}
+```
+
+## Ideas
+
+Zones are a powerful addition to the idea of simple marketplaces. By adding additional logic to approve / reject Seaport orders, many new applications are possible. Zones could potentially be used by new marketplaces built on top of Seaport to:
```
Comment (0age) [inline docs/ZoneDocumentation.md:31]: Potentially hyperbolic

File: test/index.js [hunk 16]
```diff
@@ -37,124 +33,1002 @@ const {
...
+      stubZone,
+      createOrder,
+      createMirrorBuyNowOrder,
+      createMirrorAcceptOfferOrder,
+      withBalanceChecks,
+      checkExpectedEvents,
+    } = await seaportFixture(owner));
+  });
+
+  describe("Zone - Global Pausable", async () => {
```
Comment (0age) [inline test/index.js:157]: Rename to PausableZone and move to its own test file

File: test/index.js [hunk 17]
```diff
@@ -37,124 +33,1002 @@ const {
...
+      );
+
+      // owner nukes the zone
+      gpDeployer.pause(zoneAddr);
+
+      await expect(
+        marketplaceContract.connect(buyer).fulfillOrder(order, toKey(false), {
+          value,
+        })
+      ).to.be.reverted;
```
Comment (0age) [inline test/index.js:763]: Assert the correct revert reason

**Review by 0age — CHANGES_REQUESTED** [review]
Left a number of nits on lint if warnings, documentation, and test file location / a missing assertion; once that’s all addressed LGTM

File: docs/ZoneDocumentation.md (hunk 2 above)
Comment (cupOJoseph) [reply docs/ZoneDocumentation.md:14]: this has been implemented so theres no reason to add a "could be implemented"

File: docs/ZoneDocumentation.md (hunk 15 above)
Comment (cupOJoseph) [reply docs/ZoneDocumentation.md:14]: Disagree, all of these are new use-cases.

File: contracts/zones/PausableZone.sol (hunk 6 above)
Comment (cupOJoseph) [reply contracts/zones/PausableZone.sol:249]: replaced with a passed in address argument

**Comment by ryanio** [comment]
Here's just the diff for zones-1.1 -> main: https://github.com/ProjectOpenSea/seaport/compare/main...zones-1.1 (switch to tab "Files Changed")

**Review by ryanio — APPROVED** [review]
lgtm!

**Comment by Rotemy** [comment]
Hey, I'm wondering if we can add to the Zone an option to run a static call to a given contract to verify the order before / after the sale. On Wyvern, you could have set a static target and static data to be executed after the sale. 
telegram-cloud-photo-size-4-5917799476406368425-y

That was a great way to add some sort of verification after the sale. 

The problem is that even if I want to deploy my own zone and use it, Opensea API reject the order: 

image

What do you think, can we add to the official zone a static all?

**Comment by Rotemy** [comment]
I wrote a gist explaining my request, would love to get everyone's help to get the eyes of developers from opensea on it 

https://gist.github.com/Rotemy/ce5eb81c49af78744dac63d5d087c591

**Review by 0age — APPROVED** [review]
🚀

**Pull request merged**